File: CSharp/SpaceVIL/SpaceVIL/Items/WContainer.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using SpaceVIL.Common;
using SpaceVIL.Core;

namespace SpaceVIL
{
    public delegate void EventWindowDropMethod(IItem sender, DropArgs args);

    internal sealed class WContainer : Prototype//, IWindow
    {
        public EventWindowDropMethod EventDrop;

        public override void Release()
        {
            EventDrop = null;
        }

        static int count = 0;
        internal Side _sides = 0;
        internal bool _is_fixed = false;
        private Prototype _focus = null;

        public WContainer()
        {
            SetItemName("WContainer_" + count);
            count++;
            SetStyle(DefaultsService.GetDefaultStyle(typeof(SpaceVIL.WContainer)));
            setEvents();

            // EventDrop += (sender, args) =>
            // {
            //     foreach (String path in args.Paths)
            //     {
            //         Console.WriteLine(path);
            //     }
            // };
        }

        void setEvents()
        {
            EventFocusGet += (sender) =>
            {
                GetHandler().EventFocusGet?.Invoke(sender);
            };
            EventFocusLost += (sender) =>
            {
                GetHandler().EventFocusLost?.Invoke(sender);
            };
            EventResize += (sender) =>
            {
                GetHandler().EventResize?.Invoke(sender);
            };
            EventDestroy += (sender) =>
            {
                GetHandler().EventDestroy?.Invoke(sender);
            };
            EventMouseHover += (sender, args) =>
            {
                GetHandler().EventMouseHover?.Invoke(sender, args);
            };
            EventMouseLeave += (sender, args) =>
            {
                GetHandler().EventMouseLeave?.Invoke(sender, args);
            };
            EventMouseClick += (sender, args) =>
            {
                GetHandler().EventMouseClick?.Invoke(sender, args);
            };
            EventMouseDoubleClick += (sender, args) =>
            {
                GetHandler().EventMouseDoubleClick?.Invoke(sender, args);
            };
            EventMousePress += (sender, args) =>
            {
                GetHandler().EventMousePress?.Invoke(sender, args);
            };
            EventMouseDrag += (sender, args) =>
            {
                GetHandler().EventMouseDrag?.Invoke(sender, args);
            };
            EventMouseDrop += (sender, args) =>
            {
                GetHandler().EventMouseDrop?.Invoke(sender, args);
            };
            EventScrollUp += (sender, args) =>
            {
                GetHandler().EventScrollUp?.Invoke(sender, args);
            };
            EventScrollDown += (sender, args) =>
            {
                GetHandler().EventScrollDown?.Invoke(sender, args);
            };
            EventKeyPress += (sender, args) =>
            {
                GetHandler().EventKeyPress?.Invoke(sender, args);
            };
            EventKeyRelease += (sender, args) =>
            {
                GetHandler().EventKeyRelease?.Invoke(sender, args);
            };
            EventTextInput += (sender, args) =>
            {
                GetHandler().EventTextInput?.Invoke(sender, args);
            };
            EventDrop += (sender, args) =>
            {
                GetHandler().EventDrop?.Invoke(sender, args);
            };
        }

        internal void SaveLastFocus(Prototype focused)
        {
            _focus = focused;
        }
        internal void RestoreFocus()
        {
            if (_focus != null)
                _focus.SetFocus();
            _focus = null;
        }

        protected internal override bool GetHoverVerification(float xpos, float ypos)
        {
            if (_is_fixed)
                return false;

            if (xpos > 0
                && xpos <= GetHandler().GetWidth()
                && ypos > 0
                && ypos <= GetHandler().GetHeight())
            {
                SetMouseHover(true);
            }
            else
            {
                SetMouseHover(false);
            }
            return IsMouseHover();
        }

        internal Side GetSides(float xpos, float ypos) //проблемы с глобальным курсором
        {
            if (xpos <= SpaceVILConstants.BorderCursorTolerance)
            {
                _sides |= Side.Left;
            }
            if (xpos >= GetWidth() - SpaceVILConstants.BorderCursorTolerance)
            {
                _sides |= Side.Right;
            }

            if (ypos <= SpaceVILConstants.BorderCursorTolerance)
            {
                _sides |= Side.Top;
            }
            if (ypos >= GetHeight() - SpaceVILConstants.BorderCursorTolerance)
            {
                _sides |= Side.Bottom;
            }

            return _sides;
        }
    }
}

File: spacevil_21052019/java_src/WContainer.java
package com.spvessel.spacevil;

import com.spvessel.spacevil.Core.EventDropMethodState;
import com.spvessel.spacevil.Decorations.Style;
import com.spvessel.spacevil.Flags.Side;

import java.util.*;

final class WContainer extends Prototype {
    EventDropMethodState eventDrop = new EventDropMethodState();

    @Override
    public void release() {
        eventDrop.clear();
    }

    private static int count = 0;
    List<Side> _sides = new LinkedList<>();
    boolean _is_fixed = false;
    private Prototype _focus = null;

    WContainer() {
        setItemName("WContainer_" + count);
        count++;
        // setStyle(DefaultsService.getDefaultStyle(WContainer.class));
        setStyle(Style.getWindowContainerStyle());
        setEvents();
        // eventDrop.add((sender, args) -> {
        // for (String item : args.paths) {
        // System.out.println(item);
        // }
        // });
    }

    void setEvents() {
        eventFocusGet.add((sender) -> {
            getHandler().eventFocusGet.execute(sender);
        });
        eventFocusLost.add((sender) -> {
            getHandler().eventFocusLost.execute(sender);
        });
        eventResize.add((sender) -> {
            getHandler().eventResize.execute(sender);
        });
        eventDestroy.add((sender) -> {
            getHandler().eventDestroy.execute(sender);
        });
        eventMouseHover.add((sender, args) -> {
            getHandler().eventMouseHover.execute(sender, args);
        });
        eventMouseLeave.add((sender, args) -> {
            getHandler().eventMouseLeave.execute(sender, args);
        });
        eventMouseClick.add((sender, args) -> {
            getHandler().eventMouseClick.execute(sender, args);
        });
        eventMouseDoubleClick.add((sender, args) -> {
            getHandler().eventMouseDoubleClick.execute(sender, args);
        });
        eventMousePress.add((sender, args) -> {
            getHandler().eventMousePress.execute(sender, args);
        });
        eventMouseDrag.add((sender, args) -> {
            getHandler().eventMouseDrag.execute(sender, args);
        });
        eventMouseDrop.add((sender, args) -> {
            getHandler().eventMouseDrop.execute(sender, args);
        });
        eventScrollUp.add((sender, args) -> {
            getHandler().eventScrollUp.execute(sender, args);
        });
        eventScrollDown.add((sender, args) -> {
            getHandler().eventScrollDown.execute(sender, args);
        });
        eventKeyPress.add((sender, args) -> {
            getHandler().eventKeyPress.execute(sender, args);
        });
        eventKeyRelease.add((sender, args) -> {
            getHandler().eventKeyRelease.execute(sender, args);
        });
        eventTextInput.add((sender, args) -> {
            getHandler().eventTextInput.execute(sender, args);
        });
        eventDrop.add((sender, args) -> {
            getHandler().eventDrop.execute(sender, args);
        });
    }

    void saveLastFocus(Prototype focused) {
        _focus = focused;
    }

    void restoreFocus() {
        if (_focus != null)
            _focus.setFocus();
        _focus = null;
    }

    @Override
    protected boolean getHoverVerification(float xpos, float ypos) {
        boolean result = false;
        if (_is_fixed)
            return false;

        if (xpos > 0 && xpos <= getHandler().getWidth() && ypos > 0 && ypos <= getHandler().getHeight()) {
            result = true;
            // _mouse_ptr.setPosition(xpos, ypos);
        } else {
            // _mouse_ptr.clear();
        }
        return result;
    }

    List<Side> getSides(float xpos, float ypos) {
        if (xpos <= SpaceVILConstants.borderCursorTolerance) {
            _sides.add(Side.LEFT);
        }
        if (xpos >= getWidth() - SpaceVILConstants.borderCursorTolerance) {
            _sides.add(Side.RIGHT);
        }

        if (ypos <= SpaceVILConstants.borderCursorTolerance) {
            _sides.add(Side.TOP);
        }
        if (ypos >= getHeight() - SpaceVILConstants.borderCursorTolerance) {
            _sides.add(Side.BOTTOM);
        }

        return _sides;
    }
}